Maple2.Database/Storage/Game/GameStorage.Market.cs

```diff
@@ -74,12 +163,72 @@ public partial class Request {
             return selectedResults;
         }
 
+        public ICollection<UgcMarketItem> GetUgcMarketItems(params int[] tabIds) {
+            if (tabIds.Length == 0) {
+                return Context.UgcMarketItem
+                    .Where(item => item.ListingEndTime > DateTime.Now)
+                    .AsEnumerable()
+                    .Select(ToMarketEntry)
+                    .ToList()!;
+            }
+
+            return Context.UgcMarketItem
+                .Where(item => item.ListingEndTime > DateTime.Now && tabIds.Contains(item.TabId))
+                .AsEnumerable()
+                .Select(ToMarketEntry)
+                .ToList()!;
+        }
+
+        public ICollection<UgcMarketItem> GetUgcMarketPromotedItems() {
+            ICollection<UgcMarketItem> items = Context.UgcMarketItem
+                .Where(item => item.PromotionEndTime > DateTime.Now)
+                .AsEnumerable()
+                .OrderBy(_ => Random.Shared.Next())
+                .Take(12)
+                .Select(ToMarketEntry)
+                .ToList()!;
+
+            foreach (UgcMarketItem item in items) {
+                item.Category = UgcMarketHomeCategory.Promoted;
+            }
+            return items;
+        }
+
+        public ICollection<UgcMarketItem> GetUgcMarketNewItems() {
+            ICollection<UgcMarketItem> items = Context.UgcMarketItem
+                .Where(item => item.ListingEndTime > DateTime.Now)
+                .AsEnumerable()
+                .OrderBy(item => item.CreationTime)
```

Comment:
can we do this before `AsEnumerable()`, the SQL engine should be able to evaluate it with `ORDER BY` and `LIMIT` limiting the number of entries that need to be returned from db.

Reply:
Fixed. I didn't do this for `GetUgcMarketPromotedItems` because it looks like it doesn't support Random.

Reply:
Try this:
![image](https://github.com/kOchirasu/Maple2/assets/7710340/f4a9e7bf-e6d0-4ece-ab3c-26c9294c1aa5)